docs/tutorials/add-solver.rst

```diff
@@ -0,0 +1,101 @@
+After the model has been created, parameters have been set, and the model has been discretised, the model is now a linear algebra object with the following attributes:
+
+model.concatenated_rhs
+  A :class:`pybamm.Symbol` nodes that can be evaluated at a state (``t``, ``y``) and returns the value of all the differential equations at that state, concatenated into a single vector
+model.concatenated_algebraic
+  A :class:`pybamm.Symbol` nodes that can be evaluated at a state (``t``, ``y``) and returns the value of all the algebraic equations at that state, concatenated into a single vector
+model.concatenated_initial_conditions
+  A numpy array of initial conditions for all the differential and algebraic equations, concatenated into a single vector
+model.events
+  A list of :class:`pybamm.Symbol` nodes representing events at which the solver should terminate. Specifically, the solver should terminate when any of the events in ``model.events`` evaluate to zero
+
+The role of solvers is to solve a model at a given set of time points, returning a vector of times ``t`` and a matrix of states ``y``.
+
+Base solver classes vs specific solver classes
+----------------------------------------------
+
+There is one general base solver class, :class:`pybamm.BaseSolver`, and two specialised base classes, :class:`pybamm.OdeSolver` and :class:`pybamm.DaeSolver`. The general base class simply sets up some useful solver properties such as tolerances. The specialised base classes implement a method :meth:`self.solve()` that solves a model at a given set of time points.
+
+The ``solve`` method unpacks the model, simplifies it by removing extraneous operations, (optionally) creates or calls the mass matrix and/or jacobian, and passes the appropriate attributes to another method, called ``integrate``, which does the time-stepping. The role of specific solver classes is simply to implement this ``integrate`` method for an arbitrary set of derivative function, initial conditions etc.
+
+The base DAE solver class also computes a consistent set of initial conditions for the algebraic equations, using ``model.concatenated_initial_conditions`` as an initial guess.
+
+Implementing a new solver
+-------------------------
+
+To add a new solver (e.g. My Fast DAE Solver), first create a new file (``my_fast_dae_solver.py``) in ``pybamm/solvers/``,
+with a single class that inherits from either :class:`pybamm.OdeSolver` or :class:`pybamm.DaeSolver`, depending on whether the new solver can solve DAE systems. For example:
+
+.. code-block:: python
+
+    def MyFastDaeSolver(pybamm.DaeSolver):
+
+Also add the class to ``pybamm/__init__.py``:
+
+.. code-block:: python
+
+    from .solvers.my_fast_solver import MyFastSolver
+
+You can then start implementing the solver by adding the ``integrate`` function to the class, which for an ODE solver has interface
```

Comment:
should link to the API docs for `integrate` here (see https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-python-objects), instead of reproducing the interface below. This will make it easier if we update the interface as well!